src/Core__Option.res

```diff
@@ -22,73 +22,85 @@
  * along with this program; if not, write to the Free Software
  * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */
 
+let some = value => Some(value)
+
+let flat = opt =>
+  switch opt {
+  | Some(v) => v
+  | None => None
+  }
+
 let filterU = (opt, p) =>
   switch opt {
-  | Some(x) as some if p(. x) => some
+  | Some(value) as some if p(. value) => some
   | _ => None
   }
 
-let filter = (opt, p) => filterU(opt, (. x) => p(x))
+let filter = (opt, p) => filterU(opt, (. value) => p(value))
 
 let forEachU = (opt, f) =>
   switch opt {
-  | Some(x) => f(. x)
+  | Some(value) => f(. value)
   | None => ()
   }
 
-let forEach = (opt, f) => forEachU(opt, (. x) => f(x))
+let forEach = (opt, f) => forEachU(opt, (. value) => f(value))
 
-let getExn = x =>
-  switch x {
-  | Some(x) => x
+let getExn = opt =>
+  switch opt {
+  | Some(value) => value
   | None => raise(Not_found)
   }
 
+let expect = (opt, message) =>
```

Comment:
Not sure if we shouldn't rather leave the implementation of such a function to e.g. testing frameworks (which might also want to raise a different exception).

Reply:
This isn't for testing. You should consider using this every time you reach for `getExn` because it gives you a better, easier to track down error message when the assumption that lead you to using it inevitably fails in production :upside_down_face:

Reply:
One thing that comes to mind here - do we need an additional alternative where you can pass your own exception? I recall people in the community talking about adding that themselves.

Reply:
What's the use case for that? To have a distinct exception that they can catch?

Reply:
Yeah, exactly. But let's ignore it for now, feels peripheral.

Reply:
> This isn't for testing. You should consider using this every time you reach for `getExn` because it gives you a better, easier to track down error message when the assumption that lead you to using it inevitably fails in production 🙃

You shouldn't reach for `getExn` in the first place. 😉 I rarely ever do, so I would not see myself ever needing this function.

What do you think @zth? And what did you mean here?

> I recall people in the community talking about adding that themselves.

Where was this discussed?

Reply:
Here are the two things I thought about in the community:
- https://forum.rescript-lang.org/t/the-ultimate-answer-to-belt-vs-js-in-rescript/3844/2
- https://github.com/DZakh/rescript-stdlib-vendorer/blob/main/README.md#problem-3-creating-guidelines

Both mention adding a `message` property just like @glennsl has though, so I remembered incorrectly that it was about custom exceptions rather than just a message.

As for the "feature" itself, I think it's useful (I like `expect` from Rust). I don't use `getExn` much myself either, but I know there are many that do, and I also struggle to see why you _wouldn't_ want to add a custom message in there to orient yourself when things blow up.

Reply:
Ok, right, so there is definitely demand in the community for something like this.

And what do you think about the naming? Usually we have `exn` in the method name when the function might raise. That wouldn't be the case with `expect`. Maybe it should be `getExnWithMessage`?

Reply:
I think that would ensure that absolutely no-one would pick this over `getExn`, unfortunately :wink:

Reply:
`expectExn` could be an alternative. Though I don't think the convention is that functions that raise should always use `Exn`, but rather that it signifies a variant of a function that distinguishes itself by raising. Similar to how not all functions that can return an option uses `Opt`.

Is there not an annotation that can be put on functions that raise? I thought reanalyze utilized that, but I can't find it now.

Reply:
Naming this one is hard for sure... I'm trying to come up with a good alternative as well. I like `expect` because I've used it a lot in Rust, but I'm not sure it'd feel very familiar to people who haven't used Rust. I'm ok with `getExnWithMessage`, but I understand if it feels too verbose. So let's maybe think a round or two more about what could be a good alternative name, before we decide.

@glennsl correct, it's `@raises`: https://rescript-lang.org/syntax-lookup#raises-decorator
We should eventually work our way through all the bindings and ensure all of them have `@raises` where appropriate. That'll be a lot of work though, so probably wise to do it little by little, over time. It's also less important than the general documentation is right now.

Reply:
I logged an issue for `@raises`.

Still haven't figured out a better name than `expect` or `getExnWithMessage`. Anyone else got something? Otherwise we'll need to figure out the best choice between those two.